File: src/examples.c
// examples.c
// sound examples

#include "amy.h"

// set by the arch
extern void delay_ms(uint32_t ms);

void example_reset(uint32_t start) {
    amy_event e = amy_default_event();
    e.osc = 0;
    e.reset_osc = RESET_ALL_OSCS;
    e.time = start;
    amy_add_event(&e);
}

void example_voice_alloc() {
    // alloc 2 juno voices, then try to alloc a dx7 voice on voice 0
    amy_event e = amy_default_event();
    e.patch_number = 1;
    e.voices[0] = 0;
    e.voices[1] = 1;
    amy_add_event(&e);
    delay_ms(250);

    e = amy_default_event();
    e.patch_number = 131;
    e.voices[0] = 0;
    amy_add_event(&e);
    delay_ms(250);

    // play the same note on both
    e = amy_default_event();
    e.velocity = 1;
    e.midi_note = 60;
    e.voices[0] = 0;
    amy_add_event(&e);
    delay_ms(2000);

    e = amy_default_event();
    e.velocity = 1;
    e.midi_note = 60;
    e.voices[0] = 1;
    amy_add_event(&e);
    delay_ms(2000);


    // now try to alloc voice 0 with a juno, should use oscs 0-4 again
    e = amy_default_event();
    e.patch_number = 2;
    e.voices[0] = 0;
    amy_add_event(&e);
    delay_ms(250);
}


void example_voice_chord(uint32_t start, uint16_t patch_number) {
    amy_event e = amy_default_event();
    e.time = start;
    e.patch_number = patch_number;
    e.voices[0] = 0;
    e.voices[1] = 1;
    e.voices[2] = 2;
    amy_add_event(&e);
    start += 250;

    e = amy_default_event();
    e.time = start;
    e.velocity=0.5;

    e.voices[0] = 0;
    e.midi_note = 50;
    amy_add_event(&e);
    start += 1000;

    e.voices[0] = 1;
    e.midi_note = 54;
    e.time = start;
    amy_add_event(&e);
    start += 1000;

    e.voices[0] = 2;
    e.midi_note = 56;
    e.time = start;
    amy_add_event(&e);
    start += 2000;

    e.voices[0] = 0;
    e.voices[1] = 1;
    e.voices[2] = 2;
    e.velocity = 0;
    e.time = start;
    amy_add_event(&e);
}

void example_synth_chord(uint32_t start, uint16_t patch_number) {
    // Like example_voice_chord, but use 'synth' to avoid having to keep track of voices.
    amy_event e = amy_default_event();
    e.time = start;
    e.patch_number = patch_number;
    e.num_voices = 3;
    e.synth = 0;
    amy_add_event(&e);
    start += 250;

    e = amy_default_event();
    e.velocity = 0.5;
    e.synth = 0;

    e.time = start;
    e.midi_note = 50;
    amy_add_event(&e);

    e.time += 1000;
    e.midi_note = 54;  // Will get a new voice.
    amy_add_event(&e);

    e.time += 1000;
    e.midi_note = 56;
    amy_add_event(&e);

    e.time += 2000;
    e.velocity = 0;
    // Voices are referenced only by their note, so have to turn them off individually.
    e.midi_note = 50;
    amy_add_event(&e);
    e.midi_note = 54;
    amy_add_event(&e);
    e.midi_note = 56;
    amy_add_event(&e);
}   


void example_sustain_pedal(uint32_t start, uint16_t patch_number) {
    // Reproduce TestSustainPedal, to track segfault.
    amy_event e = amy_default_event();
    e.time = start;
    e.reset_osc = RESET_ALL_OSCS;
    amy_add_event(&e);

    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.num_voices = 4;
    e.patch_number = patch_number;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 76;
    e.velocity = 1.0f;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 76;
    e.velocity = 0;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.pedal = 127;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 63;
    e.velocity = 1.0f;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 63;
    e.velocity = 0;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 67;
    e.velocity = 1.0f;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 67;
    e.velocity = 0;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 72;
    e.velocity = 1.0f;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.pedal = 0;
    amy_add_event(&e);

    start += 50;
    e = amy_default_event();
    e.time = start;
    e.synth = 1;
    e.midi_note = 72;
    e.velocity = 0;
    amy_add_event(&e);
}   



void example_patches() {
    amy_event e = amy_default_event();
    for(uint16_t i=0;i<256;i++) {
        e.patch_number = i;
        e.voices[0] = 0;
        fprintf(stderr, "sending patch %d\n", i);
        amy_add_event(&e);
        delay_ms(250);

        e = amy_default_event();
        e.voices[0] = 0;
        e.osc = 0;
        e.midi_note = 50;
        e.velocity = 0.5;
        amy_add_event(&e);

        delay_ms(1000);
        e.voices[0] = 0;
        e.velocity = 0;
        amy_add_event(&e);

        delay_ms(250);

        amy_reset_oscs();
    }
}
void example_reverb() {
    if(AMY_HAS_REVERB) {
        config_reverb(2, REVERB_DEFAULT_LIVENESS, REVERB_DEFAULT_DAMPING, REVERB_DEFAULT_XOVER_HZ); 
    }
}

void example_chorus() {
    if(AMY_HAS_CHORUS) {
        config_chorus(0.8, CHORUS_DEFAULT_MAX_DELAY, CHORUS_DEFAULT_LFO_FREQ, CHORUS_DEFAULT_MOD_DEPTH);
    }
}

// Play a KS tone
void example_ks(uint32_t start) {
    amy_event e = amy_default_event();
    e.time = start;

    e.velocity = 1;
    e.wave = KS;
    e.feedback = 0.996f;
    e.preset = 15;
    e.osc = 0;
    e.midi_note = 60;
    amy_add_event(&e);
}

// make a 440hz sine
void example_sine(uint32_t start) {
    amy_event e = amy_default_event();
    e.time = start;
    e.freq_coefs[0] = 440;
    e.wave = SINE;
    e.velocity = 1;
    amy_add_event(&e);
}

void example_multimbral_fm() {
    amy_event e0 = amy_default_event();
    amy_event e1 = amy_default_event();
    int notes[] = {60, 70, 64, 68, 72, 82};
    e1.velocity = 0.5;
    e0.patch_number = 128;
    for (unsigned int i = 0; i < sizeof(notes) / sizeof(int); ++i) {
        e1.midi_note = notes[i];
        e1.pan_coefs[0] = (i%2);
        e0.patch_number++;
        e0.voices[0] = i;
        amy_add_event(&e0);
        amy_add_event(&e1);
        delay_ms(1000);
    }
}


// Emulate the Tulip "drums()" example via event calls.
void example_drums(uint32_t start, int loops) {

    // bd, snare, hat, cow, hicow
    int oscs[] = {0, 1, 2, 3, 4};
    int presets[] = {1, 5, 0, 10, 10};

    // Reset all used oscs first, just in case
    for (unsigned int i = 0; i < sizeof(oscs) / sizeof(int); ++i) {
        amy_event e_reset = amy_default_event();
        e_reset.time = start;
        e_reset.osc = oscs[i];    
        e_reset.reset_osc = oscs[i];
        amy_add_event(&e_reset);
    }

    amy_event e = amy_default_event();
    e.time = start + 1;
    float volume = 0.5;
    e.wave = PCM;
    e.velocity = 0;

    for (unsigned int i = 0; i < sizeof(oscs) / sizeof(int); ++i) {
        e.osc = oscs[i];
        e.preset = presets[i];
        amy_add_event(&e);
    }
    // Update high cowbell.
    e = amy_default_event();
    e.time = start+1;
    e.osc = 4;
    e.midi_note = 70;
    amy_add_event(&e);

    // osc 5 : bass
    e = amy_default_event();
    e.time = start+1;
    e.osc = 5;
    e.wave = SAW_DOWN;
    e.filter_freq_coefs[COEF_CONST] = 650.0;  // LOWEST filter center frequency.
    e.filter_freq_coefs[COEF_EG0] = 2.0;  // When env0 is 1.0, filter is shifted up by 2.0 octaves (x4, so 2600.0).
    e.resonance = 5.0;
    e.filter_type = FILTER_LPF;
    strcpy(e.bp0, "0,1,500,0.2,25,0");
    amy_add_event(&e);


    const int bd = 1 << 0;
    const int snare = 1 << 1;
    const int hat = 1 << 2;
    const int cow = 1 << 3;
    const int hicow = 1 << 4;

    int pattern[] = {bd+hat, hat+hicow, bd+hat+snare, hat+cow, hat, bd+hat, snare+hat, hat};
    int bassline[] = {50, 0, 0, 0, 50, 52, 51, 0};

    e = amy_default_event();
    e.time = start+1;
    while (loops--) {
        for (unsigned int i = 0; i < sizeof(pattern) / sizeof(int); ++i) {
            e.time += 250;
            AMY_UNSET(e.freq_coefs[0]);
            AMY_UNSET(e.midi_note);
            
            int x = pattern[i];
            if(x & bd) {
                e.osc = 0;
                e.velocity = 4.0 * volume;
                amy_add_event(&e);
            }
            if(x & snare) {
                e.osc = 1;
                e.velocity = 1.5 * volume;
                amy_add_event(&e);
            }
            if(x & hat) {
                e.osc = 2;
                e.velocity = 1 * volume;
                amy_add_event(&e);
            }
            if(x & cow) {
                e.osc = 3;
                e.velocity = 1 * volume;
                amy_add_event(&e);
            }
            if(x & hicow) {
                e.osc = 4;
                e.velocity = 1 * volume;
                amy_add_event(&e);
            }

            e.osc = 5;
            if(bassline[i]>0) {
                e.velocity = 0.5 * volume;
                e.midi_note = bassline[i] - 12;
            } else {
                e.velocity = 0;
            }
            amy_add_event(&e);
        }
    }
}

void example_sequencer_drums(uint32_t start) {
    // Play a drum pattern using the low-level sequencer structure
    amy_event e = amy_default_event();
    e.time = start;
    e.reset_osc = RESET_ALL_OSCS;
    amy_add_event(&e);

    // Set tempo so 16 ticks is 108 BPM (not 48 as default).  So we make the BPM one-sixth
    e = amy_default_event();
    e.tempo = 108.0f/3;
    amy_add_event(&e);

    // Setup oscs for bd, snare, hat, cow, hicow
    int oscs[] = {0, 1, 2, 3, 4};
    int presets[] = {1, 5, 0, 10, 10};
    e = amy_default_event();
    e.time = start + 1;
    e.wave = PCM;
    for (unsigned int i = 0; i < sizeof(oscs) / sizeof(int); ++i) {
        e.osc = oscs[i];
        e.preset = presets[i];
        amy_add_event(&e);
    }
    // Update high cowbell.
    e = amy_default_event();
    e.time = start + 1;
    e.osc = 4;
    e.midi_note = 70;
    amy_add_event(&e);

    // Add patterns.
    // Hi hat every 8 ticks.
    e = amy_default_event();
    e.sequence[SEQUENCE_TAG] = 0;
    e.sequence[SEQUENCE_PERIOD] = 8;
    e.sequence[SEQUENCE_TICK] = 0;
    e.osc = 2;
    e.velocity = 1.0;
    amy_add_event(&e);

    // Bass drum every 32 ticks.
    e.sequence[SEQUENCE_TAG] = 1;
    e.sequence[SEQUENCE_PERIOD] = 32;
    e.sequence[SEQUENCE_TICK] = 0;
    e.osc = 0;
    e.velocity = 1.0;
    amy_add_event(&e);

    // Snare every 32 ticks, counterphase to BD.
    e.sequence[SEQUENCE_TAG] = 2;
    e.sequence[SEQUENCE_PERIOD] = 32;
    e.sequence[SEQUENCE_TICK] = 16;
    e.osc = 1;
    e.velocity = 1.0;
    amy_add_event(&e);

    // Cow once every other cycle.
    e.sequence[SEQUENCE_TAG] = 3;
    e.sequence[SEQUENCE_PERIOD] = 64;
    e.sequence[SEQUENCE_TICK] = 60;
    e.osc = 3;
    e.velocity = 1.0;
    amy_add_event(&e);
}

void example_sequencer_drums_synth(uint32_t start) {
    // Play a drum pattern using the low-level sequencer structure driving default system drums synth (10)
    amy_event e;

    // Add patterns.
    // Hi hat every 8 ticks.
    e = amy_default_event();
    e.sequence[SEQUENCE_TAG] = 0;
    e.sequence[SEQUENCE_PERIOD] = 24;
    e.sequence[SEQUENCE_TICK] = 0;
    e.synth = 10;
    e.midi_note = 42;  // Closed Hat
    e.velocity = 1.0;
    amy_add_event(&e);

    // Bass drum every 32 ticks.
    e.sequence[SEQUENCE_TAG] = 1;
    e.sequence[SEQUENCE_PERIOD] = 96;
    e.sequence[SEQUENCE_TICK] = 0;
    e.synth = 10;
    e.midi_note = 35;  // Std kick
    e.velocity = 1.0;
    amy_add_event(&e);

    // Snare every 32 ticks, counterphase to BD.
    e.sequence[SEQUENCE_TAG] = 2;
    e.sequence[SEQUENCE_PERIOD] = 96;
    e.sequence[SEQUENCE_TICK] = 48;
    e.synth = 10;
    e.midi_note = 38;  // Snare
    e.velocity = 1.0;
    amy_add_event(&e);

    // Cow once every other cycle.
    e.sequence[SEQUENCE_TAG] = 3;
    e.sequence[SEQUENCE_PERIOD] = 192;
    e.sequence[SEQUENCE_TICK] = 180;
    e.synth = 10;
    e.midi_note = 56;  // Cowbell
    e.velocity = 1.0;
    amy_add_event(&e);
}

void example_fm(uint32_t start) {
    // Direct construction of an FM tone, as in the documentation.
    amy_event e;

    // Modulating oscillator (op 2)
    e = amy_default_event();
    e.time = start;
    e.osc = 9;
    e.wave = SINE;
    e.ratio = 1.0f;
    e.amp_coefs[COEF_CONST] = 1.0f;
    e.amp_coefs[COEF_VEL] = 0;
    e.amp_coefs[COEF_EG0] = 0;
    amy_add_event(&e);

    // Output oscillator (op 1)
    e = amy_default_event();
    e.time = start;
    e.osc = 8;
    e.wave = SINE;
    e.ratio = 0.2f;
    e.amp_coefs[COEF_CONST] = 1.0f;
    e.amp_coefs[COEF_VEL] = 0;
    e.amp_coefs[COEF_EG0] = 1.0f;
    strcpy(e.bp0, "0,1,1000,0,0,0");
    amy_add_event(&e);

    // ALGO control oscillator
    e = amy_default_event();
    e.time = start;
    e.osc = 7;
    e.wave = ALGO;
    e.algorithm = 1;  // algo 1 has op 2 driving op 1 driving output (plus a second chain for ops 6,5,4,3).
    e.algo_source[4] = 9;
    e.algo_source[5] = 8;
    amy_add_event(&e);

    // Add a note on event.
    e = amy_default_event();
    e.time = start + 100;
    e.osc = 7;
    e.midi_note = 60;
    e.velocity = 1.0f;
    amy_add_event(&e);
}


// Minimal custom oscillator

void beeper_init(void) {
    //printf("Beeper init\n");
}

void beeper_note_on(uint16_t osc, float freq) {
    saw_down_note_on(osc, freq);
}

void beeper_note_off(uint16_t osc) {
    synth[osc]->note_off_clock = amy_global.total_blocks*AMY_BLOCK_SIZE;
}

void beeper_mod_trigger(uint16_t osc) {
    saw_down_mod_trigger(osc);
}

SAMPLE beeper_render(SAMPLE* buf, uint16_t osc) {
    return render_saw_down(buf, osc);
}

SAMPLE beeper_compute_mod(uint16_t osc) {
    return compute_mod_saw_down(osc);
}

struct custom_oscillator beeper = {
    beeper_init,
    beeper_note_on,
    beeper_note_off,
    beeper_mod_trigger,
    beeper_render,
    beeper_compute_mod
};

void example_init_custom() {
    amy_set_custom(&beeper);
}

void example_custom_beep() {
    amy_event e = amy_default_event();
    e.osc = 50;
    e.time = amy_sysclock();
    e.freq_coefs[0] = 880;
    e.wave = CUSTOM;
    e.velocity = 1;
    amy_add_event(&e);

    e.velocity = 0;
    e.time += 500;
    amy_add_event(&e);
}

void example_patch_from_events() {
    int time = amy_sysclock();
    int number = 1039;
    amy_event e = amy_default_event();
    e.time = time;
    e.patch_number = number;
    e.reset_osc = RESET_PATCH;
    amy_add_event(&e);

    e = amy_default_event();
    e.time = time;
    e.patch_number = number;
    e.osc = 0;
    e.wave = SAW_DOWN;
    e.chained_osc = 1;
    strcpy(e.bp0, "0,1,1000,0.1,200,0");
    amy_add_event(&e);

    e = amy_default_event();
    e.time = time;
    e.patch_number = number;
    e.osc = 1;
    e.wave = SINE;
    e.freq_coefs[COEF_CONST] = 131.0f;
    strcpy(e.bp0, "0,1,500,0,200,0");
    amy_add_event(&e);

    e = amy_default_event();
    e.time = time;
    e.synth = 0;
    e.num_voices = 4;
    e.patch_number = number;
    amy_add_event(&e);

    e = amy_default_event();
    e.time = time + 100;
    e.synth = 0;
    e.midi_note = 60.0f;
    e.velocity = 1.0f;
    amy_add_event(&e);

    e = amy_default_event();
    e.time = time + 300;
    e.synth = 0;
    e.midi_note = 64.0f;
    e.velocity = 1.0f;
    amy_add_event(&e);

    e = amy_default_event();
    e.time = time + 500;
    e.synth = 0;
    e.midi_note = 67.0f;
    e.velocity = 1.0f;
    amy_add_event(&e);

    e = amy_default_event();
    e.time = time + 800;
    e.synth = 0;
    e.velocity = 0.0f;
    amy_add_event(&e);
}

File: examples.py
#!/usr/bin/env python3
# examples.py
# sound examples

import amy
from time import sleep


def example_multimbral_synth():
    import example_patches as patches
    amy.send(patch_number=4, num_voices=6, synth=1)  # juno-6 preset 4 on MIDI channel 1
    amy.send(patch_number=129, num_voices=2, synth=2) # dx7 preset 1 on MIDI channel 2
    amy.send(synth=3, num_voices=4, patch_number=patches.filter_bass()) # filter bass user patch on MIDI channel 3
    amy.send(synth=4, num_voices=4, patch_number=patches.pitch_lfo()) # pitch LFO user patch on MIDI channel 4

def example_reset(start=0):
    amy.send(osc=0, reset=amy.RESET_ALL_OSCS, time=start)

def example_voice_alloc():
    # alloc 2 juno voices, then try to alloc a dx7 voice on voice 0
    amy.send(patch_number=1, voices=[0, 1])
    sleep(0.25)

    amy.send(patch_number=131, voices=[0])
    sleep(0.25)

    # play the same note on both
    amy.send(vel=1, note=60, voices=[0])
    sleep(2)

    amy.send(vel=1, note=60, voices=[1])
    sleep(2)

    # now try to alloc voice 0 with a juno, should use oscs 0-4 again
    amy.send(patch_number=2, voices=[0])
    sleep(0.25)

def example_voice_chord(patch=0):
    amy.send(patch_number=patch, voices=[0, 1, 2])
    sleep(.250)

    amy.send(vel=0.5, voices=[0], note=50)
    sleep(1)

    amy.send(vel=0.5, voices=[1], note=54)
    sleep(1)

    amy.send(vel=0.5, voices=[2], note=56)
    sleep(2)

    amy.send(vel=0, voices=[0, 1, 2])

def example_synth_chord(patch=0):
    # Like example_voice_chord, but use 'synth' to avoid having to keep track of voices
    amy.send(patch_number=patch, num_voices=3, synth=0)
    sleep(1)

    amy.send(vel=0.5, synth=0, note=50)
    sleep(1)

    amy.send(vel=0.5, synth=0, note=54)
    sleep(1)

    amy.send(vel=0.5, synth=0, note=56)
    sleep(1)

    # Voices are referenced only by their note, so have to turn them off individually
    amy.send(vel=0, synth=0, note=50)
    amy.send(vel=0, synth=0, note=54)
    amy.send(vel=0, synth=0, note=56)

def example_sustain_pedal(patch=0):
    # Reset all oscillators first
    amy.send(reset=amy.RESET_ALL_OSCS)

    amy.send(synth=1, num_voices=4, patch_number=patch)
    sleep(0.05)

    amy.send(synth=1, note=76, vel=1.0)

    sleep(0.05)
    amy.send(synth=1, note=76, vel=0)

    sleep(0.05)
    amy.send(synth=1, pedal=127)

    sleep(0.05)
    amy.send(synth=1, note=63, vel=1.0)

    sleep(0.05)
    amy.send(synth=1, note=63, vel=0)

    sleep(0.05)
    amy.send(synth=1, note=67, vel=1.0)

    sleep(0.05)
    amy.send(synth=1, note=67, vel=0)

    sleep(0.05)
    amy.send(synth=1, note=72, vel=1.0)

    sleep(0.05)
    amy.send(synth=1, pedal=0)

    sleep(0.05)
    amy.send(synth=1, note=72, vel=0)

def example_patches():
    for i in range(256):
        amy.send(patch_number=i, voices=[0])
        print(f"sending patch {i}")
        sleep(0.25)

        amy.send(voices=[0], osc=0, note=50, vel=0.5)
        sleep(1)

        amy.send(voices=[0], vel=0)
        sleep(0.25)

        amy.reset()

def example_reverb():
    amy.reverb(2, amy.REVERB_DEFAULT_LIVENESS, amy.REVERB_DEFAULT_DAMPING, amy.REVERB_DEFAULT_XOVER_HZ)

def example_chorus():
    amy.chorus(0.8, amy.CHORUS_DEFAULT_MAX_DELAY, amy.CHORUS_DEFAULT_LFO_FREQ, amy.CHORUS_DEFAULT_MOD_DEPTH)

def example_ks():
    amy.send(vel=1, wave=amy.KS, feedback=0.996, preset=15, osc=0, note=60)

def example_sine():
    amy.send(freq=[440], wave=amy.SINE, vel=1)

def example_multimbral_fm():
    notes = [60, 70, 64, 68, 72, 82]
    for i, note in enumerate(notes):
        # Two amy sends, one to load the patch, one to play it
        amy.send(voices=[i], patch_number=128+i)
        amy.send(voices=[i], note=note, vel=0.5, pan=[i*2])
        sleep(1)


def example_sequencer_drums():
    # Reset all oscillators
    amy.send(reset=amy.RESET_ALL_OSCS)

    amy.send(tempo=120.0)

    # Setup oscs for bd, snare, hat, cow, hicow
    oscs = [0, 1, 2, 3, 4]
    presets = [1, 5, 0, 10, 10]

    for osc, preset in zip(oscs, presets):
        amy.send(osc=osc, wave=amy.PCM, preset=preset)

    # Update high cowbell
    amy.send(osc=4, note=70)

    # Add patterns
    # Hi hat every 1/8th note
    amy.send(sequence=[0, 24, 0], osc=2, vel=2.0)

    # Bass drum every quarter note
    amy.send(sequence=[0, 96, 1], osc=0, vel=1.0)

    # Snare every quarter note, counterphase to BD
    amy.send(sequence=[24, 96, 2], osc=1, vel=1.0)

    # Cow once every other cycle
    amy.send(sequence=[0, 192, 3], osc=3, vel=1.0)

def example_fm():
    amy.reset()
    # Modulating oscillator (op 2)
    amy.send(osc=9, wave=amy.SINE, ratio=1.0,
            amp=[1.0, None, 0, 0])  

    # Output oscillator (op 1)
    amy.send(osc=8, wave=amy.SINE, ratio=0.2,
            amp=[1.0, None, 0, 1.0], 
            bp0="0,1,1000,0,0,0")

    # ALGO control oscillator
    amy.send(osc=7, wave=amy.ALGO, algorithm=1,
            algo_source=[0, 0, 0, 0, 9, 8])  # Only indices 4 and 5 matter here

    # Add a note on event
    sleep(.1)
    amy.send(osc=7, note=60, vel=2.0)

def example_patch_from_events():
    number = 1039

    amy.send(patch_number=number, reset=amy.RESET_PATCH)

    amy.send(patch_number=number, osc=0, wave=amy.SAW_DOWN,
            chained_osc=1, bp0="0,1,1000,0.1,200,0")

    amy.send(patch_number=number, osc=1, wave=amy.SINE,
            freq=[131.0], bp0="0,1,500,0,200,0")

    amy.send(synth=0, num_voices=4, patch_number=number)
    sleep(.1)
    amy.send(synth=0, note=60, vel=1.0)
    sleep(.3)
    amy.send(synth=0, note=64, vel=1.0)
    sleep(.2)
    amy.send(synth=0, note=67, vel=1.0)
    sleep(.3)
    amy.send(synth=0, vel=0.0) 
